File: runtime/src/utils/templateExpression.ts
export function evaluateTemplate(template: string, context: Record<string, any>): string {
  return template.replace(/\{\{([^}]+)\}\}/g, (_, expr) => {
    const trimmed = expr.trim();
    const parts = trimmed.split('|').map((p: string) => p.trim());
    const path = parts[0];
    const filter = parts[1];

    const value = resolvePath(path, context);
    if (value == null) return '';

    if (filter === 'mask') {
      const str = String(value);
      if (str.length > 6) {
        return str.slice(0, 3) + '***' + str.slice(-4);
      }
      return '***';
    }
    if (filter === 'lowercase') return String(value).toLowerCase();
    if (filter === 'uppercase') return String(value).toUpperCase();
    if (filter === 'date') return formatDate(value);

    // Auto-format ISO datetime strings
    const str = String(value);
    if (/^\d{4}-\d{2}-\d{2}T/.test(str)) {
      return formatDate(str);
    }

    return str;
  });
}

function formatDate(value: any): string {
  try {
    const d = new Date(value);
    if (isNaN(d.getTime())) return String(value);
    return d.toLocaleDateString('en-US', { month: 'short', day: 'numeric', year: 'numeric' });
  } catch {
    return String(value);
  }
}

function resolvePath(path: string, context: Record<string, any>): any {
  const segments = path.split('.');
  let current: any = context;

  for (const seg of segments) {
    if (current == null) return null;
    current = current[seg];
  }

  return current;
}

File: flutter_runtime/lib/utils/template_expression.dart
/// Evaluates template expressions like "{{record.company_name}}"
String evaluateTemplate(String template, Map<String, dynamic> record) {
  return template.replaceAllMapped(
    RegExp(r'\{\{([^}]+)\}\}'),
    (match) {
      final expr = match.group(1)!.trim();
      // Handle pipe operators like {{record.tax_id | mask}}
      final parts = expr.split('|').map((s) => s.trim()).toList();
      final path = parts[0];

      // Resolve dotted path: record.field or just field
      dynamic value;
      if (path.startsWith('record.')) {
        value = _resolvePath(record, path.substring(7));
      } else {
        value = _resolvePath(record, path);
      }

      var stringValue = value?.toString() ?? '';

      // Apply pipe transforms
      for (var i = 1; i < parts.length; i++) {
        final pipe = parts[i];
        switch (pipe) {
          case 'mask':
            // Simple mask: show first 2, mask rest
            if (stringValue.length > 4) {
              stringValue =
                  '${stringValue.substring(0, 2)}${'*' * (stringValue.length - 2)}';
            }
            break;
          case 'lowercase':
            stringValue = stringValue.toLowerCase();
            break;
          case 'uppercase':
            stringValue = stringValue.toUpperCase();
            break;
        }
      }
      return stringValue;
    },
  );
}

dynamic _resolvePath(Map<String, dynamic> data, String path) {
  final segments = path.split('.');
  dynamic current = data;
  for (final segment in segments) {
    if (current is Map<String, dynamic>) {
      current = current[segment];
    } else {
      return null;
    }
  }
  return current;
}
